### DISCUSSION TITLE
Should PyO3 fail to compile if a function name equals a reserved Python keyword?

### DISCUSSION THREAD

**Discussion by pemistahl** [PyO3/pyo3#3367] [General]
I'm pretty new to PyO3 and have just incidentally found out that PyO3 allows to compile a function or method with its name being equal to a reserved Python keyword such as `from` or `import`. Trying to call such a function from Python expectedly results in a `SyntaxError: invalid syntax`.

Shouldn't PyO3 prohibit compiling such functions or at least suggest to rename them? This isn't documented anywhere either.

Minimal Rust example to reproduce: 

```rust
#[pyclass]
struct Builder {}

#[pymethods]
impl Builder {
    #[classmethod]
    fn from(cls: &PyType) {}
}

#[pymodule]
fn foo(_py: Python<'_>, m: &PyModule) -> PyResult<()> {
    m.add_class::<Builder>()?;
    Ok(())
}
```

This fails in Python because `from` is a reserved keyword:

```python
>>> from foo import Builder
>>> Builder.from()
  File "<stdin>", line 1
    Builder.from()
                ^^^^
SyntaxError: invalid syntax
```

What is your opinion?

**Comment by davidhewitt** [comment]
I suppose `getattr(Builder, 'from')()` would work but is extremely ugly.

Warnings are hard to do well in proc macros (we have a bit of a nasty hack for deprecations). I would be open to a PR which made this an error with a `#[pyo3]` option to allow. But I'm also fine with this just as-is until we get the ability to warn nicely.

**Comment by mejrs** [comment]
A warning would be nice, but there isn't a proper way to emit warnings. Perhaps once something like https://github.com/rust-lang/rust/issues/54140 lands we should do it.
